src/engine/contactsengine.cpp

```diff
@@ -515,13 +529,27 @@ class JobThread : public QThread
         {
             QMutexLocker locker(&m_mutex);
             m_running = false;
-            m_wait.wakeOne();
         }
+        m_wait.wakeOne();
         wait();
     }
 
     void run();
 
+    bool waitForRunning()
+    {
+        QMutexLocker locker(&m_mutex);
+        if (!m_running) {
+            m_wait.wait(&m_mutex);
+        }
```

Comment:
Do you mean, the dtor may set running back to false before the caller of this function locks the mutex?  In that case, no - the thread object is only destroyed by deletion of the engine, and the main thread will have to get this mutex first.